### PR TITLE
#6662 New UseOrleansClient for generic HostBuilder

### PR DESCRIPTION
**Pull request by dpbevin**
First pass for review at utilising Generic Host for the Orleans client with a `UseOrleansClient` method. This aligns to the existing `UseOrleans` for composing a slio.

Also addresses https://github.com/dotnet/orleans/issues/7202.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=7276) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=7276) before we can accept your contribution.<br/><br/>:x: dpbevin [sign now](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=7276)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/orleans?pullRequest=7276) it.</sub>

**Comment by ReubenBond** [comment]
Thanks for doing this @dpbevin! It looks like the right direction to me.

**Comment by ReubenBond** [comment]
I rebased your PR on main and pushed some minor tweaks

**Comment by ReubenBond** [comment]
I think we should remove the Connect/Close/Abort methods on IClusterClient, too. I'm taking a look at what that would take now.

**Comment by remcoros** [comment]
Is this required to be built around IHostBuilder instead of directly on IServiceCollection?

When adding services like Mvc/AppInsights/EFCore, we can use AddXXX on IServiceCollection, without an IHostBuilder. I can imagine Orleans can do the same.

**Comment by ReubenBond** [comment]
If it's directly on IServiceCollection then you cannot access IConfiguration or anything else on the host builder context

**Comment by remcoros** [comment]
I don't think 'UseOrleansClient' needs IConfiguration or anything else from the hostbuilder context.

It feels weird to add 'application services (like orleans)' outside of 'ConfigureServices' or, in the case of asp.net, outside the 'Startup' class.

For convenience, a 'UseOrleansClient' on HostBuilder can still be added, but that just redirects to: hostBuilder.ConfigureServices(services => services.AddOrleansClient())

I can do a PR on top of this one if you'd like, as from the looks of it, this should be possible and also aligns more with the patterns of adding application services on top of IServiceCollection. (AddMvc / AddApplicationInsights / AddEfCore / etc.)

**Comment by ScarletKuro** [comment]
There must be definitely a default implementation of IHostedService  for a Client.
But do I understand correctly that this PR and if you **only** provide IHostedService means a more tight integration with GenericHost and like it will force you using only that? I'm not sure if it's a good idea for a Client API because you might want not to use IHost? It might sound weird because nowadays in .Net Core everything is a worker/asp.net core applications that uses GenericHost by default, but it still  has the right to exist?
I usually expect Client API to have something like:
`Orleans.Client` - simple library without GenericHost, just call new() pass some options in constructor, and you are good to go. Or like the current `IClientBuilder.Build()` that uses inside own container(ServiceProvider) to bootstrap.
`Orleans.Client.GenericHost` - extensions methods for `Orleans.Client` to integrate with **existing** GenericHost, IHostedService implementation etc.

**Comment by ReubenBond** [comment]
The client leverages dependency injection and has a fixed lifecycle (typically living for the duration of the application), but it shouldn't be concerned with things such as whether the process has received a SIGINT.
The current pattern requires creating multiple DI containers and therefore dual-configuring cross-cutting concerns such as logging. Aligning the client and server with a single pattern which removes those issues is preferable, I believe.

That being said, we could have a separate *client host builder* like we do today. It would likely wrap `HostBuilder`. I'm not in favor of that, since it increases the surface for misconfiguration and for things to go wrong.

I'm not against making both UseOrleans & UseOrleansClient support both `IHostBuilder` and `IServiceCollection` as long as the developer experience is good.

**Comment by ReubenBond** [comment]
cc @davidfowl

**Comment by ReubenBond** [comment]
To summarize the changes in this PR:

* Added `UseOrleansClient` on `IHostBuilder`
* Added `AddOrleansClient` on `IServiceCollection`
* Added `AddOrleans` on `IServiceCollection` (complements `UseOrleans` on `IHostBuilder`)
* `IClientBuilder` no longer has: `Build()`, `Properties`, `ConfigureHostConfiguration(...)`, `ConfigureAppConfiguration(...)`, `ConfigureServices(...)` (available as an extension method), `UseServiceProviderFactory<T>(...)`, or `ConfigureContainer<T>(...)` - it only has an `IServiceCollection Services { get; }` property.
* `ISiloBuilder` also only has a single property, `IServiceCollection Services { get; }`
* `IClusterClient` no longer has lifecycle properties (`Connect(...)`, `Close()`, `Abort()`, etc) and only has an `IServiceProvider Services { get; } property plus the methods it inherits from `IGrainFactory`: the client now uses an `IHostedService` instead

Constructing a client now looks like this:

``` C#
using var host = new HostBuilder()
  .UseOrleansClient(clientBuilder => ...)
  ...
  .Build();
host.StartAsync();
var client = host.Services.GetService<IClusterClient>();
...
```

**Comment by galvesribeiro** [comment]
Great work! It would be good to expose an `IOptions<TOption>`  that allow people to configure the retry logic for the initial connection. Similar to the `Fun that is passed to `IClusterClient.Connect([Func<Exception, Task<bool>> retryFilter = null])` Today. 

I'll finally get rid of a custom `ClusterClientService : IHostedService` which its solo purpose was to connect/disconnect the client when the `IHost` start/stop. Thanks!

:shipit:

**Comment by ScarletKuro** [comment]
I looked at changes and I here's my two cents on the subject:
Now seems like I can't make my own `IHostedService` at all since there is no Connect/Close API(which is ok, but I thought alternatives would be provided), and this is a problem because I want the client to connect only when the host is completely started `IHostApplicationLifetime.ApplicationStarted` and because `OutsideRuntimeClient` and `ClusterClientLifecycle` are internal classes there is no way for me to connect to the cluster. I kinda miss the `ClientBuilder.Build` because I could get `ClusterClient` which would allow me to make my own `IClusterClient` wrappers like LazyClusterClient
```csharp
class LazyClusterClient : IClusterClient
{
    private readonly Lazy<IClusterClient> _lazyClusterClient;

    public LazyClusterClient(Lazy<IClusterClient> lazyClusterClient)
    {
        _lazyClusterClient = lazyClusterClient;
    }
    
    ...
}

class ClusterClientHostedService : IHostedService
{
    public ClusterClientHostedService(IClientBuilder clientBuilder)
    {
        Client = new LazyClusterClient(new Lazy<IClusterClient>(clientBuilder.Build));
    }
}
```
And help me to initialize Orleans client only during connection. Why? Well I have multiple isolated clusters with different clients with own `ClientMessagingOptions/ClusterOptions/RedisClustering(connectionString)` settings, and It's not running on k8 and I have so called CentralConfig(another `IHostedService`) that will run before Orleans, download/update the config file and then Orleans would use properties from this file. 
I do agree that this changes are somehow the right direction, but my point is that the flexibility is kinda gone.

Also question, does the `ClusterClient.StartAsync`connects in background or if there is no initial cluster connection, it won't let IHost move forward until the connection is established?

**Comment by ReubenBond** [comment]
Hi @ScarletKuro, thanks for bringing up those concerns.

The `ClientBuilder` today creates its own DI container and manages all of that lifecycle. This PR pulls that container outside of the ClientBuilder's control and moves the lifecycle into an IHostedService, but you can still control these things if you want. Eg, you could create a new container and manually call into the `IHostedService` which the `ClientBuilder` registers.

Here's a rough example of how that might look:

``` C#
_clientServices = new ServiceCollection()
  .AddOrleansClient(... config ...)
  .Build();
_hostedServices = _clientServices.GetService<IEnumerable<IHostedService>>();

// Later, when you're ready to connect
foreach (var service in _hostedServices)
{
  await service.StartAsync(cancellationToken);
}

// When you want to disconnect
foreach (var service in _hostedServices)
{
  await service.StopAsync(cancellationToken);
}

// When you want to dispose the client
if (_clientServices is IAsyncDisposable asyncDisposable)
{
  await asyncDisposable.DisposeAsync();
}
else
{
  _clientServices.Dispose();
}
```
This roughly follows the pattern from the internal `Host` implementation: https://github.com/dotnet/runtime/blob/3d171f9a6e78770f66cb861de007ec91c407a803/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L63

Would that work for your case? If there was a pre-packaged way of doing this, would that be preferable? It might come in the form of a builder which wraps the above pattern and returns a new interface, perhaps named `IClientHost`, which has the relevant start/stop methods on it.

An entirely separate approach is to download that configuration you mentioned before configuring the one central `IHostBuilder` so that all configuration is available when you need it. I personally prefer that approach. If you need DI at that stage, then you can create (and later dispose) a separate container for initialization purposes. Would that work?

Please let me know if I've failed to address some aspect, too.

**Comment by galvesribeiro** [comment]
>  Is there any chance to get nuget/compiled files of this PR?

You can pull the branch locally and build the nuget.

> Btw, I think we also should not forget about Orleans.MultiClient, since this will be a breaking change, maybe we could ship this function out of the box when multiple .AddOrleansClient(builder=> {}) are added.

I was talking with @ReubenBond yesterday about that. We can't have multiple `IClusterClient` sharing the same DI container. What I believe `Orleans.MultiClient` does, is to have multiple DI containers, one per client.

**Comment by sanzor** [comment]
I have been following this thread and i wanted to ask you how should a Controller (for example) or IHostedService fetch a IClusterClient that is also connected?
Currently if i inject in a controller class the `IClusterClient`  or in the Startup using AddHostedService extension the IClusterObject is not null but it has the flag `IsInitialized=false`


**Silo**

```
IHostBuilder host = CreateHostBuilder(args);
                host.UseOrleans(builder =>
                {
                    builder
                    .UseLocalhostClustering()
                    .Configure<EndpointOptions>(opt => opt.AdvertisedIPAddress = System.Net.IPAddress.Loopback)
                    .AddLogStorageBasedLogConsistencyProvider()
                        .Configure<ClusterOptions>(options =>
                        {
                            options.ClusterId = "mycluster";
                            options.ServiceId = "myservice";
                        })
                    .ConfigureApplicationParts(parts =>
                    {
                      //  some parts
                    });
                });
                await host.RunConsoleAsync();
```

**Startup**

```
 public void ConfigureServices(IServiceCollection services)
        {
       
            services.AddHostedService(services =>
            {
                var clusterClient = services.GetService<IClusterClient>();
                //some code//
            });
        }
```

**Controller**

```
public class SomeController:ControllerBase
{
      SomeController(IClusterClient client)
      {
      }
}
```


**TLDR**

In both the controller and IHostedService the IClusterClient object is not initialized and throws exception when trying to connect to silo.
I read the docs and it said i do not need to configure the Client if it reisdes on the same process with the Silo.

**Comment by ScarletKuro** [comment]
> I read the docs and it said i do not need to configure the Client if it reisdes on the same process with the Silo.

This is correct.
Try to use `IGrainFactory` instead. Even though if everything is done correctly the `IClusterClient` should work as well(I'm actually not sure if it's true for console applications, since for asp.net core it is, but it could be different)

**Comment by sanzor** [comment]
Ok so i have tried injecting the `IGrainFactory` and again when using DI for hosted service and controller it throws saying its not yet started , has the flag `IsInitialized` `false`.

I actually have an ASP NET Web Api and not a console app. Could this be the case ? Am i starting the silo wrong ?

**Comment by ScarletKuro** [comment]
> Am i starting the silo wrong within my application?

Looks like. I made a simple Silo example co-hosted with the ASP.NET Core https://github.com/ScarletKuro/SimpleOrleansSample
In sample you will see swagger where you can execute your controller, when you execute it will use `IGrainFactory` to invoke the `IDemoGrain`.

<details>
  <summary>Some details</summary>
  
Basically I have 3 parts
1. SimpleOrleansSample.Interfaces - the grain interface, I just have one demo interface. It only references `Microsoft.Orleans.CodeGenerator.MSBuild` and `Microsoft.Orleans.Core.Abstractions`
2. SimpleOrleansSample.Grains - the grain implementation for `SimpleOrleansSample.Interfaces` that references `SimpleOrleansSample.Interfaces` project as well `Microsoft.Orleans.CodeGenerator.MSBuild` and `Microsoft.Orleans.Core.Abstractions`.
3. SimpleOrleansSample - our Silo with ASP.NET Core. References `SimpleOrleansSample.Grains` and `Microsoft.Orleans.OrleansRuntime` and important csproj settings
```XML
<EnableUnsafeBinaryFormatterSerialization>true</EnableUnsafeBinaryFormatterSerialization>
<ServerGarbageCollection>true</ServerGarbageCollection>
<ConcurrentGarbageCollection>true</ConcurrentGarbageCollection>
 ```
 and in program 
 ```csharp
 .UseOrleans(siloBuilder =>
                {
                    siloBuilder.UseLocalhostClustering()
                        .Configure<ClusterOptions>(options =>
                        {
                            options.ClusterId = "demo-gateway";
                            options.ServiceId = "demo-gateway";
                        })
                        .Configure<EndpointOptions>(options =>
                        {
                            options.AdvertisedIPAddress = IPAddress.Loopback;
                        })
                        .ConfigureApplicationParts(parts =>
                        {
                            parts.AddFromApplicationBaseDirectory();
                        });
                });
```
This is like the bare minimum you need.
  
</details>

**Comment by sanzor** [comment]
Ok so i have taken out the `RunConsoleAsync` and used the simple `Run`. However this wasn't enough it seems , the DI `IClusterClient` would yet again be not initialized.
However when using DI with the `IGrainFactory` it works !
Out of curiosity , I still  do not understand what is the difference between the two mentioned interfaces and their different behaviour ; do you by chance have any clue ?

**Comment by AntonPetrov83** [comment]
Hi!

I was going to use multiple `IClusterClient` objects to connect to clusters each having a unique ClusterId.

The idea was to have a single frontend cluster which forwards requests to backend clusters. I was going to choose target cluster and `IClusterClient` object depending on user's request version.

So now it looks like manual call to `new ClusterClientBuilder().Build()` is not an option any more?

What is an idiomatic way of implementing the described architecture?
